### ISSUE TITLE
Middleware does not run before cached content is loaded

### ISSUE THREAD

**Issue by mgoodenough** [vercel/next.js#54001]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 22.2.0: Fri Nov 11 02:08:47 PST 2022; root:xnu-8792.61.2~4/RELEASE_X86_64
    Binaries:
      Node: 16.13.2
      npm: 8.1.2
      Yarn: N/A
      pnpm: N/A
    Relevant Packages:
      next: 13.4.14-canary.5
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.1.6
    Next.js Config:
      output: N/A
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

Middleware / Edge (API routes, runtime)

### Link to the code that reproduces this issue or a replay of the bug

https://github.com/mgoodenough/nextjs-middleware-caching

### To Reproduce

**Step 1:** From the navigation menu on the Home page, click the Dashboard Link. Prior to page load, the middleware will log the following snippet to the console: pathname => /dashboard

**Step 2:** From the navigation menu on the Dashboard page, click the Home Link. Prior to page load, the middleware will log the following snippet to the console: pathname => /

**Step 3:** From the navigation menu on the Home page, click the Dashboard Link. Prior to page load, the middleware does not run.

### Describe the Bug

When navigating from one page to another in the Next.js app via the `Link` component, middleware will only run on the first visit and will not run on subsequent visits to previously viewed (cached) pages.

### Expected Behavior

When navigating from one page to another in the Next.js app via the `Link` component, middleware will run before cached content and routes are matched.

### Which browser are you using? (if relevant)

Chrome Version 114.0.5735.133

**Comment by browyn** [comment]
Any update on how to fix this ?

**Comment by quentingrchr** [comment]
Same issue here, and the reproduction repository still has the bug 👍🏻

**Comment by NickG-NZ** [comment]
Having the same issue. Is it possible to disable the cache? Even better can it be disabled just for specified 'protected' routes?

**Comment by professor-oats** [comment]
Having the same issue in my project. Anyone have any best practice workaround/solution yet?

**Comment by NickG-NZ** [comment]
Okay turns out the caching system is a big controversy with the app router.
Long story short, currently the cache cannot be disabled and lasts for 30 seconds. If you call router.refresh() it clears the cache.
Basically don't rely on middleware to protect page routes. Instead use a server component which wraps the protected page

**Comment by professor-oats** [comment]
Interesting. I have router.refresh() inside the code when handling logout (hook) so a new sign-in will be possible and update the elements correspondingly.

In my case it still loads cached elements when making a new sign-in, after the first, but now I know it can rely either on where/when the router.refresh() is used or if I can change the conditions for the displayed elements somehow.

**Comment by professor-oats** [comment]
Super thanks for your insight man. I started to fill some other parts of the code with router.refresh() after pushing a new route and now it works. MVP

**Comment by mgoodenough** [comment]
@arnaud-deprez  I agree, the middleware should kick in before the caching system per the documentation you referenced, but in my experience (thus the reason I opened this issue) that was not happening.  The middleware would not trigger on cached pages.

**Comment by arnaud-deprez** [comment]
There is [a discussion going on too](https://github.com/vercel/next.js/discussions/43675)

**Comment by accountunavaible** [comment]
can we please fix this? This is causing my client so much trouble!

**Comment by marcoshernanz** [comment]
Thanks for the report. I reproduced this on 13.4.14-canary.5 and current canary. In production, the destination is prefetched and middleware runs for that request. The later client navigation is served from the Router Cache, so there is no new server request for middleware to handle. This is the documented behavior for prefetched static routes. In the reproduction, `prefetch={false}` causes middleware to run on each navigation.

The separate stale redirect case after cookies or other request state changes is tracked in #88937. Closing this as expected Router Cache behavior.
